mcp/src/main/java/io/airlift/mcp/reference/ReferenceServerTransport.java

```diff
@@ -0,0 +1,189 @@
+import static jakarta.servlet.http.HttpServletResponse.SC_OK;
+import static jakarta.servlet.http.HttpServletResponse.SC_SERVICE_UNAVAILABLE;
+import static jakarta.ws.rs.core.HttpHeaders.ACCEPT;
+import static jakarta.ws.rs.core.MediaType.APPLICATION_JSON;
+import static jakarta.ws.rs.core.MediaType.SERVER_SENT_EVENTS;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+// based off of io.modelcontextprotocol.server.transport.HttpServletStatelessServerTransport.java
+public class ReferenceServerTransport
+        extends HttpServlet
+        implements McpStatelessServerTransport
+{
+    private static final Logger logger = Logger.get(ReferenceServerTransport.class);
+
+    private final McpJsonMapper mcpJsonMapper;
+    private final String mcpEndpoint;
+    private volatile McpStatelessServerHandler mcpHandler;
+    private volatile boolean isClosing;
+
+    @Inject
+    public ReferenceServerTransport(McpMetadata metadata, McpJsonMapper mcpJsonMapper)
+    {
+        this.mcpJsonMapper = requireNonNull(mcpJsonMapper, "jsonMapper is null");
+        this.mcpEndpoint = metadata.uriPath();
+    }
+
+    @Override
+    public void setMcpHandler(McpStatelessServerHandler mcpHandler)
+    {
+        this.mcpHandler = requireNonNull(mcpHandler, "mcpHandler is null");
+    }
+
+    @Override
+    public Mono<Void> closeGracefully()
+    {
+        return Mono.fromRunnable(() -> isClosing = true);
+    }
+
+    @Override
+    public void destroy()
+    {
+        closeGracefully().block();
+        super.destroy();
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response)
+            throws IOException
+    {
+        boolean isMcpEndpoint = request.getRequestURI().endsWith(mcpEndpoint);
+        response.sendError(isMcpEndpoint ? SC_METHOD_NOT_ALLOWED : SC_NOT_FOUND);
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest request, HttpServletResponse response)
+            throws IOException
+    {
+        String requestURI = request.getRequestURI();
+        if (!requestURI.endsWith(mcpEndpoint)) {
+            response.sendError(SC_NOT_FOUND);
+            return;
+        }
+
+        if (isClosing) {
+            response.sendError(SC_SERVICE_UNAVAILABLE, "Server is shutting down");
+            return;
+        }
+
+        McpTransportContext transportContext = McpTransportContext.create(ImmutableMap.of(CONTEXT_REQUEST_KEY, request));
+
+        String accept = request.getHeader(ACCEPT);
+        if (accept == null || !(accept.contains(APPLICATION_JSON) && accept.contains(SERVER_SENT_EVENTS))) {
+            responseError(response, SC_BAD_REQUEST, invalidRequest("Both application/json and text/event-stream required in Accept header"));
+            return;
+        }
+
+        try {
+            BufferedReader reader = request.getReader();
+            String body = reader.lines().collect(Collectors.joining("\n"));
```

Comment:
Please add `TODO: use Reader.readAllLines once on JDK 25` as a reminder for myself ;)